### ISSUE TITLE
Spacing transform changes physical extent of images

### ISSUE THREAD

**Issue by yanlend** [Project-MONAI/MONAI#5061]
**Describe the bug**
After applying the Spacing transform, the physical extent that an image covers is changing.

**To Reproduce**
```
import numpy as np
from monai.data.meta_tensor import MetaTensor
from monai.transforms import Spacing

pixdim_new = [2.0, 1.0, 0.2]
pixdim_old = [1.0, 1.0, 1.0]
affine = np.diag(pixdim_old + [1])
data = np.random.randn(1, 128, 128, 32)

img = MetaTensor(data, affine=affine).to("cpu")
res: MetaTensor = Spacing(pixdim=pixdim_new)(img)

np.testing.assert_allclose(
    np.array(img.shape[1:]) * pixdim_old,
    np.array(res.shape[1:]) * pixdim_new
)
```

**Expected behavior**
Resampled image covers same physical space as original volume, in particular if it is possible because the resampling factors are integers. In fact, during upsampling (last dimension), the physical extent is too small.
I might understand the behavior now if `align_corners` is true, in the sense that the centers of the corner pixels define the extent and not the borders.
From `align_corners` documentation: "if set to False, they are instead considered as referring to the corner points of the input’s corner pixels, making the sampling more resolution agnostic"

**Screenshots**
none

**Environment**

```
MONAI version: 0.9.1
Numpy version: 1.21.6
Pytorch version: 1.12.1+cpu
MONAI flags: HAS_EXT = False, USE_COMPILED = False, USE_META_DICT = False
MONAI rev id: 356d2d2f41b473f588899d705bbc682308cee52c

```

**Additional context**

**Comment by wyli** [comment]
Currently there is an output spatial shape parameter to override the automatically computed shape: https://github.com/Project-MONAI/MONAI/blob/4780efedf633bdcdf7cb954d1d892674bf283296/monai/transforms/spatial/array.py#L498

I don't think the rounding errors are in general avoidable, but perhaps it's a good idea to recompute the result's affine based on the output shape, similar to the `recompute_scale_factor` parameter  of [torch.nn.functional.interpolate](https://pytorch.org/docs/stable/generated/torch.nn.functional.interpolate.html#torch.nn.functional.interpolate).

**Comment by yanlend** [comment]
Thanks for the quick response.

Rounding errors are a different topic. They occur when exact sampling is not feasible. They could be handled by changing the physical extent of the image (subpixel cropping or padding) or by slightly changing the `pixdim`. In both cases, the resulting affine matrix could be recomputed to reflect the change.

The issue mentioned above occurs even without rounding. Note that the resampling factors in the image are 0.5, 1, and 5. For the given image shape, this would intuitively result in an image shape of 64, 128, 160 exactly.
The actual result has a shape of 64, 128, **156**. I should have put those numbers in the original report.

This is caused by using the center of the first/last voxels as corner in `compute_shape_offset`, which would be fine in my opinion with `align_corners = True`
However, `align_corners = False` is stating that the reshaping is based on the corners of the corner pixels. In that case, the physical extent of the image should not change (up to rounding) in my opinion, and thus give an output size of 64, 128, 160 here.

**Comment by wyli** [comment]
I see, your ask is simpler than I thought... using this patch seems to improve the align_corners=False case:
```diff
diff --git a/monai/data/utils.py b/monai/data/utils.py
index fa0e3aa4..26f0efd4 100644
--- a/monai/data/utils.py
+++ b/monai/data/utils.py
@@ -844,7 +844,7 @@ def compute_shape_offset(
     sr = len(shape)
     in_affine_ = convert_data_type(to_affine_nd(sr, in_affine), np.ndarray)[0]
     out_affine_ = convert_data_type(to_affine_nd(sr, out_affine), np.ndarray)[0]
-    in_coords = [(0.0, dim - 1.0) for dim in shape]
+    in_coords = [(-0.5, dim - 0.5) for dim in shape]
     corners: np.ndarray = np.asarray(np.meshgrid(*in_coords, indexing="ij")).reshape((len(shape), -1))
     corners = np.concatenate((corners, np.ones_like(corners[:1])))
     corners = in_affine_ @ corners
@@ -854,7 +854,7 @@ def compute_shape_offset(
         raise ValueError(f"Affine {out_affine_} is not invertible") from e
     corners_out = inv_mat @ corners
     corners_out = corners_out[:-1] / corners_out[-1]
-    out_shape = np.round(corners_out.ptp(axis=1) + 1.0)
+    out_shape = np.round(corners_out.ptp(axis=1))
     mat = inv_mat[:-1, :-1]
     k = 0
     for i in range(corners.shape[1]):
```

**Comment by yanlend** [comment]
Yes! This fixes the mentioned problem.

However, `aligned_corners=True` then does not preserve the border pixels as mentioned in the docs for nn.functional.interpolate:  If set to True, the input and output tensors are aligned by the center points of their corner pixels, preserving the values at the corner pixels. 
This fails:
```
np.testing.assert_equal(float(img[0,0,0,0]), float(res[0,0,0,0]))
```

But this check also fails with the previous implementation. Using `torch.nn.functional.interpolate` directly, the check succeeds, so there is another issue in between.

**Comment by wyli** [comment]
thanks, I submit a patch to revise the `align_corners=True` case of `Spacing` https://github.com/Project-MONAI/MONAI/pull/5075. when the flag is set to true, it's treating each voxel as a cube with a physical extent. when it's false, each voxel is considered as a point and the distance between them is the spacing:
 
![Screenshot 2022-09-05 at 10 05 17](https://user-images.githubusercontent.com/831580/188412109-b97836bf-2190-4fdc-9f13-9458450fb64a.png)

so to maintain the physical extent, we set `align_corners`  to True.

**Comment by wyli** [comment]
I think the key idea is that setting align corners to true is to treat each voxel as a cube (physical extent from -0.5*spacing to 0.5*spacing), false is as a point (distance between them is spacing)...

**Comment by yanlend** [comment]
If `align_corners` is set to false, the "corner points of the corner pixels" will be aligned in pytorch, so definitely treating the voxels as a cube.

One could define a parameter to treat voxels as cubes or points as you describe, but that should be *another parameter*, because it does not correspond to the idea of `align_corners` in pytorch.

**Comment by wyli** [comment]
you are right @yanlend, I came across scipy zoom's `grid_mode` parameter, that's a better name and it's not related to the align corners parameter... https://docs.scipy.org/doc/scipy/reference/generated/scipy.ndimage.zoom.html I'll spend more time on this and probably update the docstring to match what's implemented.

**Comment by wyli** [comment]
I add a `scale_extent` boolean parameter to choose the resampling coordinate conventions:
```
scale_extent: whether the scale is computed based on the spacing or the full extent of voxels, for example, for
    a factor of 0.5 scaling:
    option 1, "o" represents a voxel, scaling the distance between voxels::
        o--o--o
        o-----o
    option 2, each voxel has a physical extent, scaling the full voxel extent::
        | voxel 1 | voxel 2 | voxel 3 | voxel 4 |
        |      voxel 1      |      voxel 2      |
    Option 1 may reduce the number of locations the requiring interpolation. Option 2 is more resolution
    agnostic, that is, resampling coordinates depend on the scaling factor, not on the number of voxels.
    Default is False, using option 1 to compute the shape and offset.
```

in this case when the affine is computed using option 2, the `align_corners=True` should be used for SpatialResampling though, as `compute_shape_offset` option 2 already implements the shift/scaling. in this case using `align_corners=False` would incorrectly compensate the coordinates.
